// svelte-language-server/dist/src/lib/documents/Document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Document = void 0;
const utils_1 = require("../../utils");
const DocumentBase_1 = require("./DocumentBase");
const utils_2 = require("./utils");
const configLoader_1 = require("./configLoader");
/**
 * Represents a text document contains a svelte component.
 */
class Document extends DocumentBase_1.WritableDocument {
    constructor(url, content) {
        super();
        this.url = url;
        this.content = content;
        this.languageId = 'svelte';
        this.scriptInfo = null;
        this.moduleScriptInfo = null;
        this.styleInfo = null;
        this.updateDocInfo();
    }
    updateDocInfo() {
        if (!this.config || this.config.loadConfigError) {
            this.config = configLoader_1.loadConfig(this.getFilePath() || '');
        }
        const scriptTags = utils_2.extractScriptTags(this.content);
        this.scriptInfo = this.addDefaultLanguage((scriptTags === null || scriptTags === void 0 ? void 0 : scriptTags.script) || null, 'script');
        this.moduleScriptInfo = this.addDefaultLanguage((scriptTags === null || scriptTags === void 0 ? void 0 : scriptTags.moduleScript) || null, 'script');
        this.styleInfo = this.addDefaultLanguage(utils_2.extractStyleTag(this.content), 'style');
    }
    /**
     * Get text content
     */
    getText() {
        return this.content;
    }
    /**
     * Set text content and increase the document version
     */
    setText(text) {
        this.content = text;
        this.version++;
        this.updateDocInfo();
    }
    /**
     * Returns the file path if the url scheme is file
     */
    getFilePath() {
        return utils_1.urlToPath(this.url);
    }
    /**
     * Get URL file path.
     */
    getURL() {
        return this.url;
    }
    addDefaultLanguage(tagInfo, tag) {
        var _a, _b;
        if (!tagInfo) {
            return null;
        }
        const defaultLang = (_b = (_a = this.config.preprocess) === null || _a === void 0 ? void 0 : _a.defaultLanguages) === null || _b === void 0 ? void 0 : _b[tag];
        if (!tagInfo.attributes.lang && !tagInfo.attributes.type && defaultLang) {
            tagInfo.attributes.lang = defaultLang;
        }
        return tagInfo;
    }
}
exports.Document = Document;
